### ISSUE TITLE
Sorting tensors

### ISSUE THREAD

**Issue by vnkmr7620** [tensorflow/tensorflow#288]
how to sort a tensor (argsort)?

**Comment by girving** [comment]
We don't have `sort` and `argsort` ops at the moment, but would welcome contributions.

**Comment by yenchenlin1994** [comment]
Hello @girving 
I guess there is a good reason why these functions are not included in the previous version,
would you please elaborate?

**Comment by girving** [comment]
The only reason is limited resources. :)  They should clearly exist, so we'd love contributions in case we don't get to it in the near term.

**Comment by jrruijli** [comment]
What about using tf.nn.top_k, with k the number of elements in the (flattened) tensor? While possibly inefficient, it works.

**Comment by edgarriba** [comment]
@girving What's the status for sorting op?

**Comment by yaroslavvb** [comment]
@edgarriba you can sort array by using `nn.top_k` followed by `tf.gather`

**Comment by edgarriba** [comment]
@yaroslavvb are those two functions differentiable though?

**Comment by nmduc** [comment]
@edgarriba nn.top_k is an evaluation op. Evaluation ops are non differentiable so the answer is no. 
You can read a bit more here: https://www.tensorflow.org/versions/r0.11/api_docs/python/nn.html#top_k

**Comment by girving** [comment]
@nmduc @edgarriba Looks differentiable to me: https://github.com/tensorflow/tensorflow/blob/43346a1c81d503f0f764e09bb2f12e31a793a168/tensorflow/python/ops/nn_grad.py#L525

**Comment by nmduc** [comment]
@girving Good to know. Thanks.
Then the doc should be changed not to make confusion.

**Comment by mstaib** [comment]
Just chiming in to say that a GPU-friendly sort function would enable efficient projection oracles for the l1-ball and simplex (http://icml2008.cs.helsinki.fi/papers/361.pdf), which would in turn make TensorFlow useful for a wide range of large-scale constrained optimization problems beyond deep learning. This and some other efficient projection oracles (e.g. l2) could maybe be rolled into ProximalGradientDescentOptimizer long-term? (as projection is a proximal operator)

Also, issue #5719 which is a feature request for GPU-enabled topK is related. Maybe these can be handled at the same time?

**Comment by nikitakit** [comment]
+1 for adding a dedicated sorting op.

While trying to find the fastest way to sort tensors in tensorflow, I found that numpy is actually faster than using top_k. Here's the code I used to test:

```python
# assume vals is already defined
*_, last_dim = [int(x) for x in vals.get_shape()]

sorted1, _ = tf.nn.top_k(vals, k=last_dim)
sorted2 = tf.py_func(np.sort, [vals], tf.float32, stateful=False)
sorted2.set_shape(vals.get_shape())
```

In my case `sorted1` takes around 2.7 ms, while `sorted2` takes around 1.7 ms. I find it disappointing the tensorflow is actually slower than calling out to Python.

**Comment by ed-alertedh** [comment]
I'm just waiting for clearance from my employer to open source my work so far. I've been able to make use of the thrust template library included with the CUDA SDK which has a sort op which could be applied fairly easily to sort an entire tensor (I believe it internally uses a radix sort). If sorting along an axis is required then more work may be needed - I have struggled to get the thrust templates to perform well in this kind of use-case.

**Comment by girving** [comment]
@martinwicke Any obstacles to using thrust?  Having GPU sort would be great.

**Comment by ed-alertedh** [comment]
Probably worth noting that Arrayfire claims that it generally performs better than thrust [1][2]. Particularly in "row-wise" cases. The advantage of thrust is that, as far as I can tell, it already works in the current build environment without needing to add any third_party headers. So I took this as the path of least resistance. I'm sure you're constantly fighting people trying to add more and more deps...

[1] https://gist.github.com/pavanky/2407814
[2] https://arrayfire.com/benchmarking-parallel-vector-libraries/

**Comment by martinwicke** [comment]
There's no fundamental problem with, although I am less than excited to add another STL implementation. 

Can we use it instead of something else we're using in core/lib? Can we check what the effect on build times is (by adding it as a dependency in a PR)?

**Comment by ringw** [comment]
If we implemented this ourselves, we could also add an efficient median function (using something like quickselect).

**Comment by sjperkins** [comment]
Note that [NVIDIA cub](https://nvlabs.github.io/cub/) was added as a dependency in [this comment](https://github.com/tensorflow/tensorflow/commits/e073322452e41e76754314aa75d543d071003fc5/tensorflow/contrib/cmake/patches/cub/CMakeLists.txt) and contains primitives for radix sorting ints and floats. Some thought will be required to see if this can be extended to complex numbers.

**Comment by ewebxml** [comment]
Will tensorflow include a topological_sort in the future?

**Comment by ekelsen** [comment]
TopK on the GPU is now implemented using CUB's radix sort, so should be about as fast as is possible even when k == N.  CPU uses std::sort as well for k == N.  So I'm going to close this issue.

**Comment by philipperemy** [comment]
@ekelsen maybe we should a `tf.sort()` function which redirects to `TopK`.

That would be more user friendly.

**Comment by ringw** [comment]
FYI I've just added `tf.contrib.framework.sort()`, and I may be able to add `tf.contrib.framework.argsort()` in a followup.

**Comment by philipperemy** [comment]
@ringw thanks great job!

**Comment by thomaspheiffer** [comment]
Hi @ringw, did you by chance ever have an opportunity to implement `tf.contrib.framework.argsort()`?

**Comment by ringw** [comment]
@thomaspheiffer Thanks for reminding me, I've finally got my argsort wrapper working and it should be merged soon!

**Comment by imsrgadich** [comment]
@ringw thanks a lot!
